rtl/mkidaranth/src/mkidaranth/image_cuber.py

```diff
@@ -0,0 +1,498 @@
+                                    with m.If(byte_array[3] == 0b11111111):
+                                        m.d.sync += self.count_overflow_flag.eq(1)
+                                        m.d.sync += write_port11.data.eq(Cat(byte_array[0],byte_array[1],byte_array[2],byte_array[3],byte_array[4],byte_array[5],byte_array[6],byte_array[7]))
+                                with m.Else():
+                                    m.d.sync += write_port11.data.eq(Cat(byte_array[0],byte_array[1],byte_array[2],byte_array[3],byte_array[4],byte_array[5],byte_array[6],(byte_array[7]+1)[:8]))
+                                    with m.If(byte_array[7] == 0b11111111):
+                                        m.d.sync += self.count_overflow_flag.eq(1)
+                                        m.d.sync += write_port11.data.eq(Cat(byte_array[0],byte_array[1],byte_array[2],byte_array[3],byte_array[4],byte_array[5],byte_array[6],byte_array[7]))
+                            with m.Default():
+                                    m.d.sync += write_port11.data.eq(Cat(byte_array[0],byte_array[1],byte_array[2],byte_array[3],byte_array[4],byte_array[5],byte_array[6],byte_array[7]))
+
+                        m.d.sync += uploading.eq(1)
+                        m.d.sync += pixel_to_upload.eq(pixel)
+                        m.d.sync += buffered_stream.ready.eq(0)
+                
+
+                with m.If(uploading):
+                    m.d.sync += write_port11.addr.eq(pixel_to_upload[0:11])
+                    m.d.comb += read_port1.addr.eq(pixel_to_upload[0:11])
+                    m.d.sync += buffered_stream.ready.eq(0)
+
+                    m.d.sync += write_port11.en.eq(1)
+
+                    with m.If(write_port11.en == 1):
+                        m.d.sync += write_port11.en.eq(0)
+                        m.d.sync += uploading.eq(0)
+                        m.d.sync += buffered_stream.ready.eq(1)
+
+
+                with m.If(i >= self.cycles_per_frame-1):
+                    m.d.sync += i.eq(0)
+                    m.d.sync += write_port11.en.eq(0)
+                    m.next = "Stalling"
+
+
+            with m.State("Stalling"):
+                with m.If(self.generate_cubes == 1):
+                    m.d.sync += i.eq(i+1)
+
+                m.d.comb += read_port1.addr.eq(self.mem_read_addr)
+                m.d.comb += self.mem_read_data.eq(read_port1.data)
+                m.d.comb += self.read_en.eq(1)
+
+                with m.If(i == self.cycles_per_frame - 1):
+                    m.d.sync += i.eq(0)
+                    m.d.comb += self.read_en.eq(0)
+                    with m.If(self.generate_cubes == 1):
+                        m.next = "Clearing"
+                    with m.Else():
+                        m.next = "Configuring"
+
+                
+        #State machine memory 2
```

Comment:
Collapse the state machine to reduce code duplication, either to a module or a for loop